src/provider/p_key.c

```diff
@@ -0,0 +1,493 @@
+/*
+ * Copyright [2021] International Business Machines Corp.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#include <stdio.h>
+#include <stdarg.h>
+#include <stddef.h>
+#include <err.h>
+#include <strings.h>
+#include <string.h>
+#include <errno.h>
+
+#include <openssl/evp.h>
+#include <openssl/core.h>
+#include <openssl/core_dispatch.h>
+#include <openssl/core_names.h>
+#include <openssl/params.h>
+
+#include "p_ibmca.h"
+
+void ibmca_keymgmt_upref(struct ibmca_key *key)
+{
+    ibmca_debug_key(key, "key: %p", key);
+
+    __sync_add_and_fetch(&key->ref_count, 1);
+
+    ibmca_debug_key(key, "ref_count: %u", key->ref_count);
+}
+
+unsigned int ibmca_keymgmt_downref(struct ibmca_key *key)
+{
+    ibmca_debug_key(key, "key: %p ", key);
+
+     if (key->ref_count > 0)
```

Comment:
To be more specific:  The problem is that the value of the ref_count might be changed by another thread before we return from this function and decide based on the return value what to do.  The whole path from ref count decrement to the actual decision to delete the object has to be atomic to remove above race condition.

Reply:
You are right, the return of the new value is the real problem.

How about this:
```
unsigned int ibmca_keymgmt_downref(struct ibmca_key *key)
{
    const struct ibmca_prov_ctx *provctx = key->provctx;
    unsigned int cnt;

    ibmca_debug_key(key, "key: %p ", key);

    if (key->ref_count == 0)
        return 0;

    cnt = __sync_sub_and_fetch(&key->ref_count, 1);
    ibmca_debug_ctx(provctx, "ref_count: %u", cnt);

    return cnt;
}
```

With this, if 2 threads have a reference to the key (i.e. ref_count = 2) and then concurrently free the key, both find out that ref_counf is > 0 (no matter when they check), and then decrement the ref_count atomic. Each one returns its new value: One of it returns 1 and the other returns 0. Those that returned 0 will actually free the object. The other will not touch the key object anymore (not even for tracing).

What is still not thread save is the check for ref_count already being zero, but this really is a case where an error has happend before already. One must not have a reference to a key whos ref_count is zero. If you don't have a reference, you won't call downref(). Any change of the ref_count (which was zero) by another thread (which also would not be allowed to have a reference to the key with a zero ref_count) would be incorrect in itself.

Reply:
Well, the `if (key->ref_count == 0)` case is anyway the case where somebody already violated our calling convention.  But the `return 0` would lead to a free on a key object that has already been freed and we do not get anything.  At least some debug message (maybe also an assertion?) and a non-0 return value such that we do not make things worse?  I simply do not like code paths that lead to double frees (even if someone violates our calling convention).

Reply:
Good point about the return 0..... How about this:
```
unsigned int ibmca_keymgmt_downref(struct ibmca_key *key)
{
    const struct ibmca_prov_ctx *provctx;
    unsigned int cnt;

    assert(key->ref_count > 0);

    provctx = key->provctx;
    ibmca_debug_ctx(provctx, "key: %p ", key);

    cnt = __sync_sub_and_fetch(&key->ref_count, 1);
    ibmca_debug_ctx(provctx, "ref_count: %u", cnt);

    return cnt;
}
```
It will abort with an assert if the ref_count is zero, and does not touch the key before the assertion. It still does not check for key being non-NULL, but that would not help much I guess anyway. If the key has already been freed, then the dereferencing of key to check the ref_count of course can already cause a segfault or similar....

Reply:
The assert is only available if `NDEBUG` is not defined.  But that is fine for me.  It is anyway just a goody to try to figure out that we are already screwed...

Reply:
right. It could also be questioned if a library is supposed to do an abort (as part of the assert), but as you said, its just a try to make things not worse that they already are when the assert is hit.